File: 이것이 코딩테스트다/05.이진 탐색/5.떡볶이 떡 만들기.py
n, m = 4, 6
data = [19, 15, 10, 17]

start = 0
end = max(data)
result = 0

while start <= end:
  mid = (start + end) // 2
  sum_data = 0
  for i in data:
    if i > mid:
      sum_data += i - mid
  # 떡양이 부족한 경우 -> 조사하는 값을 줄여야함
  if sum_data < m:
    end = mid - 1
  else:
    result = mid
    start = mid + 1
   
print(result)

File: 이것이 코딩테스트다/05.이진 탐색/5.떡볶이 떡 만들기.js
const texts = [
  [4, 6],
  [19, 15, 10, 17],
];

const [n, m] = texts.shift();
const data = texts.pop();

let start = 0;
let end = Math.max(...data);

while (start <= end) {
  let mid = Math.floor((start + end) / 2);

  let total = 0;
  for (const d of data) {
    if (d > mid) {
      total += d - mid;
    }
  }

  if (total < m) {
    end = mid - 1;
  } else {
    result = mid;
    start = mid + 1;
  }
}
console.log(result);
